### ISSUE TITLE
Add a compiler optimization to use unordered operations when possible

### ISSUE THREAD

**Issue by ronawho** [chapel-lang/chapel#12150]
For codes like RA-atomics we have a normal, and unordered (currently called buffered) variants. These looks like:

```chpl
forall (_, r) in zip(Updates, RAStream()) do
  T(r & indexMask).xor(r);
```

```chpl
forall (_, r) in zip(Updates, RAStream()) do
  T(r & indexMask).xorBuff(r); // eventually `T(r & indexMask).xor(r, order=unordered)`
flushAtomicBuff(); // won't be needed after #12065
```

The unordered operations have up to a [5X performance improvement](https://chapel-lang.org/perf/16-node-xc/?startdate=2018/11/28&enddate=2019/01/24&configs=gnuugniqthreads&graphs=hpccbufferedraatomicsperfgupsn233,balehistogramperfmbspernode,baleindexgatherperfmbspernode):

![image](https://user-images.githubusercontent.com/1588337/51716745-f38a8900-1ff2-11e9-8837-f8d81583ea53.png)

For cases like RA (and bale indexgather and bale histogram) the compiler should be able to automatically turn normal blocking operations into these unordered operations. In cases like RA we're only doing a "write" and there's no reads within the same loop iteration and by using a forall the users is asserting that there are no ordering constraints.

Possibly relevant: https://github.com/chapel-lang/chapel/issues/9782#issuecomment-397370514

**Comment by mppf** [comment]
Some notes about the legality of the transformation:

``` chapel
forall r in rindex {
  A[r].add(1);
}

A starting point is a can-transform-last-statement-in-forall rule, which happens to match the existing patterns. This rule would translate the add to addBuff because it's the last statement in the forall body.


What about a loop like this:
``` chapel
forall r in rindex {
  A[r].add(1);
  B[r].add(1);
}
```

Here I think if A and B can't alias, both of these could be buffered atomics, because these atomic operations are write only.

``` chapel
forall r in rindex {
    A[r].add(1);
    while B[r].read() == 0  {  ...   }
}
```

Here the add can't be changed to addBuff because another task could wait for the add and update B[r].

So, here my current understanding of the applicability of this optimization for atomics:
 * if the atomic write (e.g. add) is last statement in forall, can always optimize including last statement in each branch in a conditional
 * if it's earlier, later statements can't rely on / use the result or be atomic/sync operations that read anything (because of possibilty for another task to update an atomic based on the write)

For a PUT or GET, a later atomic in the loop body with release order would need to prevent the optimization (since another task might be waiting on the memory to be update after the atomic notifies it, e.g.).

**Comment by mppf** [comment]
``` chapel
forall r in rindex {
  A[r].add(1);
  B[r].add(1);
}
```

I'm having second thoughts about this loop. Isn't it the case that by making both of them unordered, we'd relax the sequential consistency rule? That doesn't matter for this code, but there could conceivably be some other task working with the arrays A and B at the same time.

What if it were:

``` chapel
forall r in rindex {
  A[r].add(1);
  B[r].add(100000);
}
```

couldn't a task looking at A[r] and B[r] take an action that assumes that if B[r] is > 100000, then A[r] > 1, for example?

I think the transformation would still be valid if it were relaxed atomics, though:

``` chapel
forall r in rindex {
  A[r].add(1, order=relaxed);
  B[r].add(100000, order=relaxed);
}
```

**Comment by mppf** [comment]
I wonder about a pattern like this:

``` chapel
forall ... {
  takeLock(someLock); // maybe a compare-and-swap / test-and-set loop
  someLock.write(0); // releases the lock
}
```

Supposing that we could translate this into a program using the existing unordered operation support in the runtime, it might look like this:
``` chapel
forall ... {
  while (myLock.compareExchange(0, 1, order=memory_order_acquire)) do chpl_task_yield();
  .... do something under the lock ...
  myLock.sub(1, order=memory_order_release); // releases the lock
}
```

Now, there are two questions:
 1. Would this program deadlock, or run to completion, with the current unordered support in the runtime?
 2. Would the performance of this program be much slower if the myLock.sub call is translated into an unordered operation?

**Comment by mppf** [comment]
Maybe it's sufficient to check for the lock case by avoiding the optimization if the forall contains a loop containing an atomic operation.

**Comment by ronawho** [comment]
Another case to consider:

We automatically flush unordered ops on task termination, but it's possible for a forall to run serially and not create any tasks. We probably need to insert a call to `chpl_comm_unordered_task_fence()` after the forall loop. (That function doesn't exist yet, so in the interim you could call `chpl_comm_task_end()`)

**Comment by mppf** [comment]
Here is a program that "fails" when optimized with the new optimization:

``` chapel
config const n = 10000;
config const maxTask = 10000;

use ChapelTaskID;

pragma "fn synchronization free"
extern proc chpl_task_getId():chpl_taskID_t;

proc main() {
  var iterationsForTask:[0..maxTask] atomic int;

  on Locales[numLocales-1] {
    forall i in 1..n {
      var id = chpl_task_getId():int;
      if 0 <= id && id <= maxTask {
        iterationsForTask[id].add(1);
        assert(iterationsForTask[id].read() == 1);
        // do something useful...
        iterationsForTask[id].sub(1);
      }
    }
  }
}
```

The program is checking that each task "runs" only one forall iteration at a time. It's notable that it would fail with vectorization as well (and yes, some vectorization techniques can handle atomic ops). Which makes me think that the task for this one is to make sure the language specification indicates that forall loop bodies might appear to run as simultaneously running multiple iterations per task.

**Comment by mppf** [comment]
@ronawho convinced me that we'd like to support task/thread local storage idioms in forall loops, and as a result the above program should work.

Here is an more plausible example that @ronawho created:

``` chapel
record buffer {
  var b: [1..64] int;
  var cnt: int;

  proc enqueue(int i) {
    b[cnt] = i;
    cnt += 1;
    if cnt == b.size {
      flushBuffer(b);
      cnt = 0;
    }
  }
}

var perTaskBuff: [0..maxTask] buffer;

forall i in 1..n with (var id=taskCounter.fetchAdd(1)) {
  perTaskBuff[id].enqueue(i);
}
``

Here, the concern is that if a single task could work on two forall iterations at once, then they fill in the same b slot in the buffer.

Another way to put that is with this sketch:
``` chapel
forall i in 1..n with (var id=taskCounter.fetchAdd(1)) {
  ref tls = ...;
  tls.b[tls.cnt] = i;
  tls.cnt += 1;
}
```

If that were translated into this:
``` chapel
forall i in 1..n by 2 with (var id=taskCounter.fetchAdd(1)) {
  ref tls = ...;
  tls.b[tls.cnt] = i;
  tls.b[tls.cnt] = i;
  tls.cnt += 1;
  tls.cnt += 1;
}
```

we'd miss updates, even assuming each of the above statements are atomic.

Here's another example:
```
forall i in 1..n with (var id=taskCounter.fetchAdd(1)) {
  ref tls = ...;
  tls.b[tls.cnt] = i;
  tls.cnt += 1;
  if tls.cnt == size {
    flushSomething();
    tls.cnt = 0; // last statement
  }
}
```

Here one might think `tls.cnt = 0` is the last statement in the forall, therefore it can be put off until the end of the task.

But this isn't true. Because the task is using task-local storage and needs the tls.cnt operation to be completed before the next iteration because otherwise the task-local storage ends up in an inconsistent state.

Here is another variant of the above:
``` chapel
forall i in 1..n with (var id=taskCounter.fetchAdd(1)) {
  ref tls = ...;
  tls.b[tls.cnt] = i;
  if tls.cnt == size-1 {
    flushSomething();
    tls.cnt = -1;
  }
  var tmp = tls.cnt + 1;
  tls.cnt = tmp; // last statement
}
```

Lastly, the task-local-storage issue would extend even to optimizing atomics. Here is an example that *defers* some atomic operation on TLS. (Note that atomics on TLS are reasonable and come up in things like work-stealing). This example doesn't have any blocking in it, either.

``` chapel
var TlsArray:[1..ntasks] tls;

forall i in 1..n {
  ref tls = TlsArray[task_id];
  if tls.deferredSignal then
    tls.signal.add(1);

  tls.deferredSignal = 1;
  tls.happensBeforeSignal.add(1, order=memory_order_relaxed);
}
tls deinit:
  if tls.deferredSignal then
    tls.signal.add(1);

in some other task:
wait on tls.signal == 1
  assert(tls.happensBeforeSignal >= 1);
```

**Comment by mppf** [comment]
Note that the legality of this transformation is connected to the legality of vectorization, see #7761.

**Comment by mppf** [comment]
I have some questions about legality that we should try to address before turning it on by default:

1. What exactly are the properties of `forall` it is relying on? We should make sure the language specification clearly describes these properties. (Can a forall include a barrier? Can the loop body rely on some ordering property among the iterations?)
2. Lifetime checking is (by design) incomplete. If it has an incomplete answer, we don't want the optimization to create incorrect programs. Can we identify the cases that aren't handled correctly today and could lead to misoptimization?
3. Is the identification of task-local-storage uses sufficient? What about forall loops that use knowledge of the parallel iterator to know which iterations go in which tasks? What if the parallel iterator yields task numbers? What if we add more mechanisms to get something like a "task id" (for GPU support, e.g.)? Would it be better to check for the issue in terms of a write/read relationship data between iterations?

**Comment by ronawho** [comment]
After chatting offline Michael and I think: 
 - (1) and (3) are largely documentation tasks. We need to more clearly document the semantics of a forall loop that we're relying on (such as order-independence.) That work is captured in https://github.com/Cray/chapel-private/issues/174
 - For (2) -- I couldn't come up with any cases that wouldn't also require synchronization. If you resize an array while iterating over it in a forall loop the only way to do that would be with synchronization and if you did something like clear an owned class that has an array during a forall loop you couldn't then have a statement following it that modifies the array value, so we're current safe because of the last statement restriction of the optimization.

**Comment by ronawho** [comment]
Optimization added in https://github.com/chapel-lang/chapel/pull/12269 and enabled by default in https://github.com/chapel-lang/chapel/pull/12952. Follow on work captured in https://github.com/Cray/chapel-private/issues/174. 


Here's a pretty graph showing that RA with no source code changes is now 4.5x faster:

![image](https://user-images.githubusercontent.com/1588337/57322337-29421200-70b8-11e9-978b-95a3ee3f381f.png)
